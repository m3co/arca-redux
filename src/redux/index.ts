export * from './selectors';

export { createArcaRedux } from './store';
