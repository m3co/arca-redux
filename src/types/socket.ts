import { State, Model, Row } from './state';

export interface Socket {
  socket: SocketIOClient.Socket;
  subscribe: (Target: keyof State['Source']) => void;
  select: (Source: keyof State['Source'], PK?: Model['PK']) => void;
  delete: (Source: keyof State['Source'], PK: Model['PK']) => void;
  update: (Source: keyof State['Source'], Row: Row, PK: Model['PK']) => void;
  insert: (Source: keyof State['Source'], Row: Row) => void;
}
