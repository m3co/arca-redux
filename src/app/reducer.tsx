
import { ArcaState, ArcaActions, initialState } from './types';

export function ArcaReducer(state: ArcaState = initialState, action: ArcaActions): ArcaState {
    switch (action.type) {
        case 'Connect':
            return { ...state, active: true }
        case 'Disconnect':
            return { ...state, active: false }
        case 'Select':
            return { ...state, Sources: {
                AAU: {
                        Rows: action.Result
                    }
                }
            }
        default:
            return state
    }
}
