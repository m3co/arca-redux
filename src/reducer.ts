
import { Action, State } from './types';
import { reducers } from './reducer-models';

export const initialState: State = {
  Connected: false,
  Source: {
    'Projects': {
      Rows: [],
      Requests: [],
    },
    'Contractors': {
      Rows: [],
      Requests: [],
    },
    'FACAD-Reports': {
      Rows: [],
      Requests: [],
    },
    'FACAD-ReportFilters': {
      Rows: [],
      Requests: [],
    },
    'FACAD-CFT-AAU': {
      Rows: [],
      Requests: [],
    },
    'FACAD-CFT-Filters-AAU': {
      Rows: [],
      Requests: [],
    },
    'FACAD-preCFT-AAU': {
      Rows: [],
      Requests: [],
    },
    'FACAD-preCFT-AAU-Key': {
      Rows: [],
      Requests: [],
    },
    'FACAD-ParamsBIC': {
      Rows: [],
      Requests: [],
    },
    'FACAD-BuiltInCategories': {
      Rows: [],
      Requests: [],
    },
    'AAU': {
      Rows: [],
      Requests: [],
    },
    'AAU-QTO': {
      Rows: [],
      Requests: [],
    },
    'AAU-Tasks-Gantt': {
      Rows: [],
      Requests: [],
    },
    'AAU-APU-Tasks-Gantt': {
      Rows: [],
      Requests: [],
    },
    'AAU-Concretize': {
      Rows: [],
      Requests: [],
    },
    'APU': {
      Rows: [],
      Requests: [],
    },
    'APU-Import-Supplies': {
      Rows: [],
      Requests: [],
    },
    'APU-MetaSupplies': {
      Rows: [],
      Requests: [],
    },
    'APU-P-Supplies': {
      Rows: [],
      Requests: [],
    },
    'APU-QTO': {
      Rows: [],
      Requests: [],
    },
    'APU-Tasks': {
      Rows: [],
      Requests: [],
    },
    'APU-Tasks-Gantt': {
      Rows: [],
      Requests: [],
    },
    'APU-Assign': {
      Rows: [],
      Requests: [],
    },
    'AEU': {
      Rows: [],
      Requests: [],
    },
    'Budget-AAU': {
      Rows: [],
      Requests: [],
    },
    'Budget-AAU-vs-General': {
      Rows: [],
      Requests: [],
    },
    'Tasks-Month-CashFlow-AAU': {
      Rows: [],
      Requests: [],
    }
  },
};

export const reducer = (state: State = initialState, action: Action): State => {
  switch (action.type) {
    case 'Connect':
      return {
        ...state,
        Connected: true
      };
    case 'Subscribe':
      return {
        ...state,
        Source: {
          ...state.Source,
          [action.Source]: {
            ...state.Source[action.Source],
            Subscribed: action.Subscribed,
          }
        }
      };
    case 'GetInfo':
      return {
        ...state,
        Source: {
          ...state.Source,
          [action.Source]: {
            ...state.Source[action.Source],
            Info: action.Result,
          }
        }
      };
    case 'Select':
      return {
        ...state,
        Source: {
          ...state.Source,
          [action.Source]: {
            ...state.Source[action.Source],
            Rows: action.Result,
          }
        }
      };
    case 'Insert':
    case 'Delete':
    case 'Update':
      return {
        ...state,
        Source: {
          ...state.Source,
          [action.Source]: {
            ...state.Source[action.Source],
            Requests: (action.Success) ?
              [...state.Source[action.Source].Requests].filter(
                (ID: string): boolean => ID !== action.ID) :
              [...state.Source[action.Source].Requests, action.ID],
          }
        }
      };
    case 'insert':
      const { Row } = action;
      return {
        ...state,
        Source: {
          ...state.Source,
          [action.Source]: {
            ...state.Source[action.Source],
            Rows: [Row, ...state.Source[action.Source].Rows]
          }
        }
      };
    case 'update':
      return {
        ...state,
        Source: {
          ...state.Source,
          [action.Source]: {
            ...state.Source[action.Source],
            Rows: reducers[action.Source].Update(state, action.Row, action.PK)
          }
        }
      };
    case 'delete':
      return {
        ...state,
        Source: {
          ...state.Source,
          [action.Source]: {
            ...state.Source[action.Source],
            Rows: reducers[action.Source].Delete(state, action.Row, action.PK)
          }
        }
      };
    default:
      return state;
  }
};
