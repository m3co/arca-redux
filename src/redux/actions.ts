import { arcaDataType } from './types';
import { ARCA_DATA } from './constant';

export const setArcaData = (arcaData: string): arcaDataType => ({
  type: ARCA_DATA,
  arcaData,
});
