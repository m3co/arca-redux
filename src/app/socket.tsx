
import * as Socket from 'socket.io-client';
import { v4 as uuid4 } from 'uuid';
import { ArcaState, ArcaActions, ArcaResponses, AAURow, AAUPK } from './types';
import { Store } from 'redux'

export class ArcaSocket {
  private io: SocketIOClient.Socket;

  public constructor(store: Store<ArcaState, ArcaActions>) {
    const io = this.io = Socket();

    io.on('connect', (): void => {
      store.dispatch({
        type: 'Connect'
      });
    });

    io.on('jsonrpc', (response: ArcaResponses): void => {
      switch (response.Method) {
        case 'Subscribe':
          break;
        case 'Unsubscribe':
          break;
        case 'Update':
          console.log(response, 'by update'); // Still don't know how to notify about this response... need to talk with @NovikovAntonY
          break;
        case 'insert':
        case 'delete':
        case 'update':
          if (response.Context.Target === 'AAU') {
            store.dispatch({
              type: 'Notify',
              Context: {
                Target: response.Context.Target
              },
              Method: response.Method,
              Row: response.Row,
              PK: {
                Key: response.Row.Key
              }
            });
          }
          break;
        case 'Select':
          store.dispatch({
            type: response.Method,
            Context: {
              Source: response.Context.Source
            },
            Result: response.Result
          });
          break;
        case 'GetInfo':
          store.dispatch({
            type: response.Method,
            Context: {
              Source: response.Context.Source
            },
            Result: response.Result
          });
          break;
        default:
          console.log(response);
          break;
      }
    });
  }

  public Select(Table: string): void {
    this.io.emit('jsonrpc', {
      ID: uuid4(),
      Context: {
        Source: Table
      },
      Method: 'Select'
    });
  }

  public Update(Table: string, Row: AAURow, PK: AAUPK): void {
    this.io.emit('jsonrpc', {
      ID: uuid4(),
      Context: {
        Source: Table
      },
      Method: 'Update',
      Params: {
        Row, PK
      }
    });
  }

  public Subscribe(Table: string): void {
    this.io.emit('jsonrpc', {
      ID: uuid4(),
      Method: 'Subscribe',
      Params: {
        Target: Table,
      },
    });
  }

  public GetInfo(Table: string): void {
    this.io.emit('jsonrpc', {
      ID: uuid4(),
      Method: 'GetInfo',
      Context: {
        Source: Table,
      },
    });
  }
}
