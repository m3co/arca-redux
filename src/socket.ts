
import Socket from 'socket.io-client';
import { v4 as uuid4 } from 'uuid';
import { Store } from 'redux';

import { Field, Model, Row, State } from './types';
import { handleResponse } from './socket-handleResponse';

export class ARCASocket {
  public readonly store: Store<State>;
  private readonly io: SocketIOClient.Socket;

  public constructor(store: Store, io = Socket()) {
    this.store = store;
    this.io = io;
    this.io.on('connect', (): void => {
      this.store.dispatch({
        type: 'Connect'
      });
    });

    this.io.on('jsonrpc', handleResponse(store));
  }

  public GetInfo = (Source: keyof State["Source"]): void => {
    this.io.emit('jsonrpc', {
      ID: uuid4(),
      Method: 'GetInfo',
      Context: {
        Source
      },
    });
  }

  public Select = (Source: keyof State["Source"]): void => {
    this.io.emit('jsonrpc', {
      ID: uuid4(),
      Method: 'Select',
      Context: {
        Source
      },
    });
  }

  public Subscribe = (Target: keyof State["Source"]): void => {
    this.io.emit('jsonrpc', {
      ID: uuid4(),
      Method: 'Subscribe',
      Params: {
        Target
      },
    });
  }

  public Delete = (Source: keyof State["Source"], row: Row, ID?: string): void => {
    const state = this.store.getState();
    const { Info } = state.Source[Source];
    if (Info) {
      const columnsPK = Info.Fields
        .filter((field: Field): boolean => field.Primary)
        .map((field: Field): keyof Model["PK"] => field.Name as keyof Model["PK"]);
      const PK = columnsPK.reduce((acc: Model["PK"], column) => {
        acc[column] = row[column];
        return acc;
      }, {} as Model["PK"]);
      const request = {
        ID: ID || uuid4(),
        Context: {
          Source
        },
        Method: 'Delete',
        Params: {
          PK
        }
      };
      this.store.dispatch({
        Source,
        type: 'Delete',
        ID: request.ID,
      });
      this.io.emit('jsonrpc', request);
    }
  }

  public Update = (Source: keyof State["Source"], row: Row, ID?: string): void => {
    const state = this.store.getState();
    const { Info } = state.Source[Source];
    if (Info) {
      const columnsPK = Info.Fields
        .filter((field: Field): boolean => field.Primary)
        .map((field: Field): keyof Model["PK"] => field.Name as keyof Model["PK"]);
      const PK = columnsPK.reduce((acc: Model["PK"], column) => {
        acc[column] = row[column];
        return acc;
      }, {} as Model["PK"]);
      const request = {
        ID: ID || uuid4(),
        Context: {
          Source
        },
        Method: 'Update',
        Params: {
          Row: row,
          PK: PK,
        }
      };
      this.store.dispatch({
        Source,
        type: 'Update',
        ID: request.ID,
      });
      this.io.emit('jsonrpc', request);
    }
  }

  public Insert = (Source: keyof State["Source"], row: Row, ID?: string): string => {
    const request = {
      ID: ID || uuid4(),
      Context: {
        Source
      },
      Method: 'Insert',
      Params: {
        Row: row
      }
    }
    this.io.emit('jsonrpc', request);
    this.store.dispatch({
      Source,
      type: 'Insert',
      ID: request.ID,
    });
    return request.ID;
  }
}
