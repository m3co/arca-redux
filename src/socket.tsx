
import * as Socket from 'socket.io-client';
import { v4 as uuid4 } from 'uuid';
import { ArcaState, ArcaActions, ArcaResponses, ArcaEntries,
  AAU, FACADParamsBIC, FACADSchedules } from './types';
import { Store } from 'redux'

export class ArcaSocket {
  private io: SocketIOClient.Socket;

  public constructor(store: Store<ArcaState, ArcaActions>) {
    const io = this.io = Socket();

    io.on('connect', (): void => {
      store.dispatch({
        type: 'Connect'
      });
    });

    io.on('jsonrpc', (response: ArcaResponses): void => {
      switch (response.Method) {
        case 'Subscribe':
          break;
        case 'Unsubscribe':
          break;
        case 'Update':
          console.log(response, 'by update'); // Still don't know how to notify about this response... need to talk with @NovikovAntonY
          break;
        case 'insert':
        case 'delete':
        case 'update':
          switch (response.Context.Target) {
            case 'AAU':
              store.dispatch({
                type: 'Notify',
                Context: {
                  Target: response.Context.Target
                },
                Method: response.Method,
                Row: response.Row as AAU["Row"]
              });
              break;
            case 'FACAD-ParamsBIC':
              store.dispatch({
                type: 'Notify',
                Context: {
                  Target: response.Context.Target
                },
                Method: response.Method,
                Row: response.Row as FACADParamsBIC["Row"]
              });
              break;
            case 'FACAD-Schedules':
              store.dispatch({
                type: 'Notify',
                Context: {
                  Target: response.Context.Target
                },
                Method: response.Method,
                Row: response.Row as FACADSchedules["Row"]
              });
              break;
            default:
              break;
          }
          break;
        case 'Select':
          switch (response.Context.Source) {
            case 'AAU':
              store.dispatch({
                type: 'Select',
                Context: {
                  Source: response.Context.Source
                },
                Result: response.Result as AAU["Row"][]
              });
              break;
            case 'FACAD-ParamsBIC':
              store.dispatch({
                type: 'Select',
                Context: {
                  Source: response.Context.Source
                },
                Result: response.Result as FACADParamsBIC["Row"][]
              });
              break;
            case 'FACAD-Schedules':
              store.dispatch({
                type: 'Select',
                Context: {
                  Source: response.Context.Source
                },
                Result: response.Result as FACADSchedules["Row"][]
              });
              break;
            default:
              break;
          }
          break;
        case 'GetInfo':
          store.dispatch({
            type: response.Method,
            Context: {
              Source: response.Context.Source
            },
            Result: response.Result
          });
          break;
        default:
          console.log(response);
          break;
      }
    });
  }

  public Select(Source: string): void {
    this.io.emit('jsonrpc', {
      ID: uuid4(),
      Context: {
        Source
      },
      Method: 'Select'
    });
  }

  public Insert(Source: string, Row: ArcaEntries["Row"]): void {
    this.io.emit('jsonrpc', {
      ID: uuid4(),
      Context: {
        Source
      },
      Method: 'Insert',
      Params: {
        Row
      }
    });
  }

  public Delete(Source: string, PK: ArcaEntries["PK"]): void {
    this.io.emit('jsonrpc', {
      ID: uuid4(),
      Context: {
        Source
      },
      Method: 'Delete',
      Params: {
        PK
      }
    });
  }

  public Update(Source: string, Entry: ArcaEntries): void {
    this.io.emit('jsonrpc', {
      ID: uuid4(),
      Context: {
        Source
      },
      Method: 'Update',
      Params: {
        Row: Entry.Row,
        PK: Entry.PK
      }
    });
  }

  public Subscribe(Target: string): void {
    this.io.emit('jsonrpc', {
      ID: uuid4(),
      Method: 'Subscribe',
      Params: {
        Target
      },
    });
  }

  public GetInfo(Source: string): void {
    this.io.emit('jsonrpc', {
      ID: uuid4(),
      Method: 'GetInfo',
      Context: {
        Source
      },
    });
  }
}
