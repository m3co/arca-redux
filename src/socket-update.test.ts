
import Socket from 'socket.io-client';
import { ARCASocket } from './socket';
import { reducer } from './reducer';
import { createStore } from 'redux';

const URL = 'http://localhost:8086';
const source = 'FACAD-Reports';

test(`Perform Update at ${source} over a Row`, (done): void => {
  const io = Socket(URL);
  const store = createStore(reducer);
  const socket = new ARCASocket(store, io);

  socket.Select(source);
  socket.Subscribe(source);
  socket.GetInfo(source);

  let updating = false;
  let i = 0;
  store.subscribe((): void => {
    const state = store.getState();
    const Source = state.Source[source];
    if (state.Connected && !updating &&
      Source.Info &&
      Source.Rows.length &&
      Source.Subscribed) {
      expect(Source).toMatchSnapshot();
      i++;

      updating = true;
      socket.Update(source, {
        BuiltInCategory: 'INVALID',
        ID: 19,
        Name: 'modified name',
        PathName: '12121sdgcw',
        Field1: '',
        Field2: '',
        Field3: '',
        ReportType: '',
      }, { ID:'3eff0c6b-e248-41bc-bc8f-d5233f881e1c' });
    } else if (state.Connected &&
      Source.Info &&
      Source.Rows.length &&
      Source.Subscribed &&
      Source.Requests.length) {
      expect(Source).toMatchSnapshot();
      i++;
    } else if (state.Connected &&
      Source.Info &&
      Source.Rows.length &&
      Source.Subscribed) {
      if (Source.Rows[0].ID === 19) {
        expect(Source.Rows).toMatchSnapshot();
        expect(Source.Requests.length).toBe(0);
        done();
        return;
      }
      i++;
    }
    if (i === 4) {
      fail();
    }
  });
});

test(`Perform Update at FACAD-preCFT-AAU-Key over a Row with different Row and PK`, (done): void => {
  const source = 'FACAD-preCFT-AAU-Key';
  const io = Socket(URL);
  const store = createStore(reducer);
  const socket = new ARCASocket(store, io);

  socket.Select(source);
  socket.Subscribe(source);
  socket.GetInfo(source);

  let updating = false;
  let i = 0;
  store.subscribe((): void => {
    const state = store.getState();
    const Source = state.Source[source];
    if (state.Connected && !updating &&
      Source.Info &&
      Source.Rows.length &&
      Source.Subscribed) {
      expect(Source).toMatchSnapshot();
      i++;

      updating = true;
      socket.Update(source, {
        Family: 'Concrete',
        Type: 'Concreto Muro de contencion',
        Key: '1.3.1'
      }, { ID:'61bb4f24-a89d-40d8-970a-f393dd3eb207', PK: {
        Family: 'Concrete',
        Type: 'Concreto Muro de contencion',
        Key: null
      }});
    } else if (state.Connected &&
      Source.Info &&
      Source.Rows.length &&
      Source.Subscribed &&
      Source.Requests.length) {
      expect(Source).toMatchSnapshot();
      i++;
    } else if (state.Connected &&
      Source.Info &&
      Source.Rows.length &&
      Source.Subscribed) {
      if (Source.Rows[0].Key === '1.3.1') {
        expect(Source.Rows).toMatchSnapshot();
        expect(Source.Requests.length).toBe(0);
        done();
        return
      }
      i++;
    }
    if (i === 4) {
      fail();
    }
  });
});
