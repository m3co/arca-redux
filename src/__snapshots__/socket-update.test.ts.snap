// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Perform Update at FACAD-Reports over a Row 1`] = `
Object {
  "Info": Object {
    "Actions": Object {
      "Delete": true,
      "Insert": true,
      "Update": true,
    },
    "Fields": Array [
      Object {
        "Combobox": null,
        "Editable": false,
        "Name": "ID",
        "Primary": true,
        "Required": false,
        "Select": null,
        "Type": "Number",
      },
      Object {
        "Combobox": null,
        "Editable": true,
        "Name": "Name",
        "Primary": false,
        "Required": true,
        "Select": null,
        "Type": "String",
      },
      Object {
        "Combobox": null,
        "Editable": true,
        "Name": "BuiltInCategory",
        "Primary": false,
        "Required": true,
        "Select": null,
        "Type": "String",
      },
      Object {
        "Combobox": null,
        "Editable": true,
        "Name": "PathName",
        "Primary": false,
        "Required": true,
        "Select": null,
        "Type": "String",
      },
    ],
  },
  "Requests": Array [],
  "Rows": Array [
    Object {
      "BuiltInCategory": "INVALID",
      "ID": 19,
      "Name": "1ssssxdd",
      "PathName": "12121sdgcw",
    },
    Object {
      "BuiltInCategory": "INVALID",
      "ID": 22,
      "Name": "csdcs",
      "PathName": "sdsd",
    },
  ],
  "Subscribed": true,
}
`;

exports[`Perform Update at FACAD-Reports over a Row 2`] = `
Object {
  "Info": Object {
    "Actions": Object {
      "Delete": true,
      "Insert": true,
      "Update": true,
    },
    "Fields": Array [
      Object {
        "Combobox": null,
        "Editable": false,
        "Name": "ID",
        "Primary": true,
        "Required": false,
        "Select": null,
        "Type": "Number",
      },
      Object {
        "Combobox": null,
        "Editable": true,
        "Name": "Name",
        "Primary": false,
        "Required": true,
        "Select": null,
        "Type": "String",
      },
      Object {
        "Combobox": null,
        "Editable": true,
        "Name": "BuiltInCategory",
        "Primary": false,
        "Required": true,
        "Select": null,
        "Type": "String",
      },
      Object {
        "Combobox": null,
        "Editable": true,
        "Name": "PathName",
        "Primary": false,
        "Required": true,
        "Select": null,
        "Type": "String",
      },
    ],
  },
  "Requests": Array [
    "3eff0c6b-e248-41bc-bc8f-d5233f881e1c",
  ],
  "Rows": Array [
    Object {
      "BuiltInCategory": "INVALID",
      "ID": 19,
      "Name": "1ssssxdd",
      "PathName": "12121sdgcw",
    },
    Object {
      "BuiltInCategory": "INVALID",
      "ID": 22,
      "Name": "csdcs",
      "PathName": "sdsd",
    },
  ],
  "Subscribed": true,
}
`;

exports[`Perform Update at FACAD-Reports over a Row 3`] = `
Array [
  Object {
    "BuiltInCategory": "INVALID",
    "ID": 19,
    "Name": "1ssssxdd",
    "PathName": "12121sdgcw",
  },
  Object {
    "BuiltInCategory": "INVALID",
    "ID": 22,
    "Name": "csdcs",
    "PathName": "sdsd",
  },
]
`;

exports[`Perform Update at FACAD-preCFT-AAU-Key over a Row with different Row and PK 1`] = `
Array [
  Object {
    "BuiltInCategory": "INVALID",
    "ID": 19,
    "Name": "modified name",
    "PathName": "12121sdgcw",
  },
  Object {
    "BuiltInCategory": "INVALID",
    "ID": 22,
    "Name": "csdcs",
    "PathName": "sdsd",
  },
]
`;

exports[`Perform Update at FACAD-preCFT-AAU-Key over a Row with different Row and PK 2`] = `
Object {
  "Info": Object {
    "Actions": Object {
      "Delete": false,
      "Insert": false,
      "Update": true,
    },
    "Fields": Array [
      Object {
        "Combobox": null,
        "Editable": true,
        "Name": "Family",
        "Primary": true,
        "Required": true,
        "Select": null,
        "Type": "Text",
      },
      Object {
        "Combobox": null,
        "Editable": true,
        "Name": "Type",
        "Primary": true,
        "Required": true,
        "Select": null,
        "Type": "Text",
      },
      Object {
        "Combobox": Object {
          "Display": "Description",
          "Params": Object {
            "Key": "Key",
          },
          "Source": "AAU",
          "Value": "Key",
        },
        "Editable": true,
        "Name": "Key",
        "Primary": true,
        "Required": true,
        "Select": null,
        "Type": "Text",
      },
    ],
  },
  "Requests": Array [],
  "Rows": Array [
    Object {
      "Family": "Concrete",
      "Key": null,
      "Type": "Concreto Muro de contencion",
    },
    Object {
      "Family": "Concrete",
      "Key": null,
      "Type": "enchape pared ceramica 25x35",
    },
    Object {
      "Family": "Concrete-Round-Column_16110 (1)",
      "Key": null,
      "Type": "pilote d=50",
    },
  ],
  "Subscribed": true,
}
`;

exports[`Perform Update at FACAD-preCFT-AAU-Key over a Row with different Row and PK 3`] = `
Object {
  "Info": Object {
    "Actions": Object {
      "Delete": false,
      "Insert": false,
      "Update": true,
    },
    "Fields": Array [
      Object {
        "Combobox": null,
        "Editable": true,
        "Name": "Family",
        "Primary": true,
        "Required": true,
        "Select": null,
        "Type": "Text",
      },
      Object {
        "Combobox": null,
        "Editable": true,
        "Name": "Type",
        "Primary": true,
        "Required": true,
        "Select": null,
        "Type": "Text",
      },
      Object {
        "Combobox": Object {
          "Display": "Description",
          "Params": Object {
            "Key": "Key",
          },
          "Source": "AAU",
          "Value": "Key",
        },
        "Editable": true,
        "Name": "Key",
        "Primary": true,
        "Required": true,
        "Select": null,
        "Type": "Text",
      },
    ],
  },
  "Requests": Array [
    "61bb4f24-a89d-40d8-970a-f393dd3eb207",
  ],
  "Rows": Array [
    Object {
      "Family": "Concrete",
      "Key": null,
      "Type": "Concreto Muro de contencion",
    },
    Object {
      "Family": "Concrete",
      "Key": null,
      "Type": "enchape pared ceramica 25x35",
    },
    Object {
      "Family": "Concrete-Round-Column_16110 (1)",
      "Key": null,
      "Type": "pilote d=50",
    },
  ],
  "Subscribed": true,
}
`;

exports[`Perform Update at FACAD-preCFT-AAU-Key over a Row with different Row and PK 4`] = `
Array [
  Object {
    "Family": "Concrete",
    "Key": "1.3.1",
    "Type": "Concreto Muro de contencion",
  },
  Object {
    "Family": "Concrete",
    "Key": null,
    "Type": "enchape pared ceramica 25x35",
  },
  Object {
    "Family": "Concrete-Round-Column_16110 (1)",
    "Key": null,
    "Type": "pilote d=50",
  },
]
`;
